File: js/endless-binary/dec2bin-1.js
// taidalab Version 1.1.0
// https://github.com/taidalog/taidalab
// Copyright (c) 2022 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
function checkAnswerd2b1 (answer, last_answers) {
    const hintArea = document.getElementById('hintArea');
    const errorArea = document.getElementById('errorArea');
    errorArea.innerHTML = '';
    
    const numberInput = document.getElementById('numberInput');
    const bin = escapeHtml(numberInput.value);
    console.log(bin);
    
    const quotientsAndRemainders = repeatDivision(answer, 2);
    console.log(quotientsAndRemainders);
    
    const powerOfTwos = devideIntoPowerOfTwo(answer);
    console.log(powerOfTwos);
    
    if (bin == '') {
        errorArea.innerHTML = '<span class="warning">' + answer + ' の2進法表記を入力してください。</span>';
    } else if (testBinaryString(bin) == false) {
        errorArea.innerHTML = '<span class="warning">"' + bin + '" は2進数ではありません。使えるのは半角の 0 と 1 のみです。</span>';
    } else {
        
        const binaryDigit = 8;
        const destinationRadix = 2;
        const zeroPaddedBin = bin.padStart(binaryDigit, '0');
        const taggedBin = colorLeadingZero(zeroPaddedBin);
        const dec = parseInt(bin, destinationRadix);
        console.log(taggedBin);
        console.log(dec);
        
        let historyClassName = ''
        if (dec == answer) {
            historyClassName = 'history-correct'
        } else {
            historyClassName = 'history-wrong'
        }
        
        const decimalDigit = 3;
        const spacePaddedDec = dec.toString().padStart(decimalDigit, ' ').replace(' ', '&nbsp;');
        
        const sourceRadix = 10;
        const outputArea = document.getElementById('outputArea');
        const msg1 = '<span class ="' + historyClassName + '">' + taggedBin + '<sub>(' + destinationRadix + ')</sub> = ' + spacePaddedDec + '<sub>(' + sourceRadix + ')</sub></span>';
        const msg2 = concatinateStrings(msg1, outputArea.innerHTML);
        outputArea.innerHTML = msg2;
        console.log(msg1);
        console.log(msg2);
        
        if (dec == answer) {
            let nextNumber = 0;
            let nextBin = 0;
            
            console.log(last_answers);
            do {

                do {
                    nextNumber = getRandomBetween(0, 192);
                    nextBin = nextNumber.toString(2);
                } while (countOneBit(nextBin) != 2);
                console.log(nextNumber);
                console.log(last_answers.some((element) => element == nextNumber));
            } while (last_answers.some((element) => element == nextNumber));
            
            const quotientsAndRemainders = repeatDivision(nextNumber, 2);
            console.log(quotientsAndRemainders);
            
            const powerOfTwos = devideIntoPowerOfTwo(nextNumber);
            console.log(powerOfTwos);

            const nextHint = newHintDec2Bin(nextNumber, quotientsAndRemainders, powerOfTwos);
            console.log(nextHint);
            
            document.getElementById('questionSpan').innerText = nextNumber;
            hintArea.innerHTML = nextHint;
            
            numberInput.value = '';

            const answersToKeep = 10;
            const lastAnswers = [nextNumber].concat(last_answers).slice(0, answersToKeep);
            document.getElementById('submitButton').onclick = function() { checkAnswerd2b1(nextNumber, lastAnswers); return false; };
        }
    }
    
    numberInput.focus();
}


function countOneBit (binaryString) {
    const regex = /1/g;
    const res = binaryString.match(regex);
    if (res == null) {
        return 0
    } else {
        return res.length
    }
}


function devideIntoPowerOfTwo (number) {
    function getMaxPowerOfTwo (number) {
        const indexNumber = Math.floor(Math.log(number) / Math.log(2));
        return Math.pow(2, indexNumber);
    }

    function loop (acc, number) {
        if (number == 0) {
            return acc;
        } else if (number == 1) {
            return acc.concat([1]);
        } else {
            const max = getMaxPowerOfTwo(number);
            return loop(acc.concat([max]), number - max);
        }
    }

    return loop([], number);
}


function repeatDivision (dividend, divisor) {
    const quotient = parseInt(dividend / divisor);
    const remainder = dividend - (quotient * divisor);
    if (quotient < divisor) {
        return [[quotient, remainder]];
    } else {
        return [[quotient, remainder]].concat(repeatDivision(quotient, divisor));
    }
}


function newColumnAddition (quotients_and_remainders) {
    const lengthMinusOne = quotients_and_remainders.length - 1;
    return quotients_and_remainders.slice(0, lengthMinusOne).reduceRight(
        (prev, curr) => '2<span class="column-addition-row">' + curr[0].toString().padStart(3, ' ').replace(' ', '&nbsp;') + '</span>...' + curr[1] + '<br>' + prev,
        '<span class="column-addition-row-last">' + quotients_and_remainders[lengthMinusOne][0].toString().padStart(5, ' ').replace(/ /g, '&nbsp;') + '</span>...' + quotients_and_remainders[lengthMinusOne][1]
        );
}


function newHintDec2Bin (number, quotients_and_remainders, power_of_twos) {
    return '<details><summary>ヒント: </summary>' + '<h2>考え方 1</h2>' + newHintRepeatDivision(number, quotients_and_remainders) + '<h2>考え方 2</h2>' + newHintRepeatAddition(number, power_of_twos) + '</details>'
}


function newHintRepeatDivision (number, quotients_and_remainders) {
    const firstRow = '2<span class="column-addition-row">' + number.toString().padStart(3, ' ').replace(' ', '&nbsp;') + '</span>';
    const columnAddition = newColumnAddition(quotients_and_remainders);

    const msg01 = '<div class="history-indented">';
    const msg02 = '<p>10進数を、商が 1 になるまで 2 で割り続けます。<br>';
    const msg03 = 'この時、余りを商の右に書いておきます。<br>';
    const msg04 = '商と余りを下から順に繋げると、2進数になります。</p>';
    const msg05 = '</div>';
    const msg06 = '<div class="history-indented column-addition-area">';
    const msg07 = '</div>';

    return msg01 + msg02 + msg03 + msg04 + msg05 + msg06 + firstRow + '<br>' + columnAddition + msg07
}


function newHintRepeatAddition (number, power_of_twos) {
    const hintFormat01 = '<p class="history-indented">';
    const hintFormat02 = '{0}<sub>(10)</sub> 以下で最大の2の累乗は {1}<sub>(10)</sub><br>';
    const hintFormat03 = '{0}<sub>(10)</sub> - {1}<sub>(10)</sub> = {2}<sub>(10)</sub><br>';
    const hintFormat04 = '{2}<sub>(10)</sub> 以下で最大の2の累乗は {3}<sub>(10)</sub><br>';
    const hintFormat05 = '{2}<sub>(10)</sub> - {3}<sub>(10)</sub> = {4}<sub>(10)</sub><br>';
    const hintFormat06 = 'よって、{0}<sub>(10)</sub> = {1}<sub>(10)</sub> + {3}<sub>(10)</sub><br>';
    const hintFormat07 = 'または、{0}<sub>(10)</sub> = 2<sup>{5}</sup><sub>(10)</sub> + 2<sup>{6}</sup><sub>(10)</sub>';
    const hintFormat08 = '</p>';
    const hintFormat = hintFormat01 + hintFormat02 + hintFormat03 + hintFormat04 + hintFormat05 + hintFormat06 + hintFormat07 + hintFormat08;
    const hint = formatString(hintFormat, [number, power_of_twos[0], number - power_of_twos[0], power_of_twos[1], number - power_of_twos[0] - power_of_twos[1], Math.floor(Math.log(power_of_twos[0]) / Math.log(2)), Math.floor(Math.log(power_of_twos[1]) / Math.log(2))]);
    console.log(hint);
    return hint
}


function initDec2Bin1 () {
    // initialization
    let initNumber = 0;
    let initBin = '';

    do {
        initNumber = getRandomBetween(0, 192);
        initBin = initNumber.toString(2);
    } while (countOneBit(initBin) != 2);

    const quotientsAndRemainders = repeatDivision(initNumber, 2);
    const powerOfTwos = devideIntoPowerOfTwo(initNumber);
    console.log(initNumber);
    console.log(quotientsAndRemainders);
    console.log(powerOfTwos);

    const sourceRadix = 10;
    const destinationRadix = 2;

    document.getElementById('questionSpan').innerHTML = initNumber;
    document.getElementById('srcRadix').innerHTML = '(' + sourceRadix + ')';
    document.getElementById('dstRadix').innerHTML = destinationRadix;
    document.getElementById('binaryRadix').innerHTML = '<sub>(' + destinationRadix + ')</sub>';
    document.getElementById('hintArea').innerHTML = newHintDec2Bin(initNumber, quotientsAndRemainders, powerOfTwos);
    document.getElementById('submitButton').onclick = function() { checkAnswerd2b1(initNumber, [initNumber]); return false; };
}

File: src/endless-binary/Dec2Bin1.fs
// taidalab
// https://github.com/taidalog/taidalab
// Copyright (c) 2022-2025 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
namespace Taidalab

open System
open Browser.Dom
open Browser.Types
open Taidalab.Number
open Taidalab.Text
open Taidalab.EndlessBinary
open Fermata
open Fermata.RadixConversion

module EndlessBinary =
    module Dec2Bin1 =
        let help =
            """
            10進数から2進数への変換をエンドレスで練習できます。<br>
            出題範囲は n (0&le;n&le;255) で、2の累乗の数同士の和になっています。<br>
            ヒント付きなので、考え方も身に付けられます。
            """

        let devideIntoPowerOfTwo (number: int) =
            let getMaxPowerOfTwo (number: int) = number |> log2 |> int |> pown 2

            let rec loop acc number =
                match number with
                | 0 -> acc
                | 1 -> acc @ [ 1 ]
                | _ ->
                    let max = getMaxPowerOfTwo number
                    loop (acc @ [ max ]) (number - max)

            loop [] number


        let rec repeatDivision (dividend: int) (divisor: int) : (int * int) list =
            let quotient, remainder = Math.DivRem(dividend, divisor)

            if quotient < divisor then
                [ (quotient, remainder) ]
            else
                [ (quotient, remainder) ] @ repeatDivision quotient divisor

        let newArrowBin fontSize lineCount stroke fill =
            Svg.newArrow
                (fontSize |> double |> (fun x -> x / 2. * 4.))
                (lineCount |> (fun x -> fontSize * (x - 1) + 6) |> double)
                (fontSize |> double |> (fun x -> x / 2. * 3.))
                (lineCount |> double |> (fun x -> 17.85 * x - 35.) |> (*) -1.)
                -48.
                (17.85 * (lineCount |> double) - 15.)
                (lineCount - 1 |> delayMs |> (+) 1500)
                stroke
                fill


        let newHintAnimation divisor num fontSize =
            let divRems =
                (numOpt divisor num) :: (divRemOpt divisor (repeatDivision num divisor))

            divRems
            |> List.mapi (fun i (a, b, c, d) ->
                Option.map // divisor
                    (fun x ->
                        Svg.text
                            0
                            (fontSize * (i + 1))
                            0.
                            (sprintf
                                "%d%s"
                                x
                                (Svg.animateOpacity
                                    (i |> delayMs |> (fun x -> if i = 0 then x + 1000 else x + 2000))
                                    500)))
                    a,
                Option.map // line
                    (fun x ->
                        Svg.path
                            (sprintf
                                "M %d,%d q %d,%f 0,%f h %f"
                                (fontSize / 2 + 2)
                                (fontSize * i + 6)
                                (fontSize / 2)
                                (double fontSize * 0.4)
                                (double fontSize * 0.8)
                                (double fontSize / 2. * 4.8))
                            "#000000"
                            1
                            "none"
                            0.
                            (Svg.animateOpacity (i |> delayMs |> (fun x -> if i = 0 then x + 500 else x + 1500)) 500))
                    b,
                Option.map // dividend
                    (fun x ->
                        Svg.text
                            (fontSize / 2 * 2)
                            (fontSize * (i + 1))
                            0.
                            (sprintf
                                "%s%s"
                                (x |> string |> Fermata.String.padLeft 3 ' ' |> escapeSpace)
                                (Svg.animateOpacity (i |> delayMs) 500)))
                    c,
                Option.map // remainder
                    (fun x ->
                        Svg.text
                            (fontSize / 2 * 6)
                            (fontSize * (i + 1))
                            0.
                            (sprintf "…%d%s" x (Svg.animateOpacity (i |> delayMs |> (+) 500) 500)))
                    d)
            |> List.map (fun (a, b, c, d) ->
                sprintf
                    "%s%s%s%s"
                    (Option.defaultValue "" a)
                    (Option.defaultValue "" b)
                    (Option.defaultValue "" c)
                    (Option.defaultValue "" d))
            |> List.fold
                (fun x y -> sprintf "%s%s" x y)
                (newArrowBin fontSize (List.length divRems) "#191970" "#b0e0e6")
            |> (Svg.frame (fontSize / 2 * 10) (divRems |> List.length |> (fun x -> fontSize * (x + 1))))

        let newHintRepeatDivision divisor number =
            $"""
            <div class="history-indented">
                <p>
                    10進法で表現した数を2進法で表現しなおすには、<br>
                    10進法の数を、商が 1 になるまで 2 で割り続けます。<br>
                    この時、余りを商の右に書いておきます。<br>
                    商と余りを下から順に繋げると、2進法の数になります。<br>
                    ※この下の筆算をクリックすると動きます。
                </p>
            </div>
            <div id="hint1" class="history-indented mono">
                %s{(newHintAnimation divisor number 20)}
            </div>
            """

        let newHintRepeatAddition (dec: Dec) (power_of_twos: int list) : string =
            let additionDec = power_of_twos |> List.map string |> String.concat " + "

            let additionIndex =
                power_of_twos
                |> List.map (log2 >> truncate >> int)
                |> List.map (sprintf "2<sup>%d</sup>")
                |> String.concat " + "

            let additionBin =
                power_of_twos
                |> List.map (Dec.Valid >> Dec.toBin)
                |> List.map (function
                    | Bin.Valid v -> v
                    | Bin.Invalid _ -> "")
                |> List.map (fun x -> $"%s{x}<sub>(2)</sub>")
                |> String.concat " + "

            let bin: string =
                dec
                |> Dec.toBin
                |> function
                    | Bin.Valid v -> v
                    | Bin.Invalid _ -> "-1"

            let number: int =
                dec
                |> function
                    | Dec.Valid v -> v
                    | Dec.Invalid _ -> -1

            $"""
            <p class="history-indented">
                10進法で表現した数を2進法で表現しなおすには、<br>
            </p>
            <p class="history-indented">
                <ol style="padding-left: 4rem;">
                    <li>10進法の数を「2<sup>n</sup> の数同士の足し算」に変換して、</li>
                    <li>それぞれの 2<sup>n</sup> の数を2進法で表し、</li>
                    <li>足し合わせる</li>
                </ol>
            </p>
            <p class="history-indented">
                という方法もあります。
            </p>
            <p class="history-indented">
                %d{number}<sub>(10)</sub> を 2<sup>n</sup> の数同士の足し算に変換すると
            </p>
            <p class="history-indented hint-bgcolor-gray mono">
                &nbsp;&nbsp;%s{additionDec}<br>
                = %s{additionIndex}
            </p>
            <p class="history-indented">
                になります。<br>
            </p>
            <p class="history-indented">
                次に、それぞれの 2<sup>n</sup> の数を2進法で表します。<br>
                2<sup>n</sup> の数を2進法で表すには、1 の後に 0 を n 個続けます。<br>
                そのため、%s{additionIndex} は2進法で
            </p>
            <p class="history-indented hint-bgcolor-gray mono">
                &nbsp;&nbsp;%s{additionBin}<br>
            </p>
            <p class="history-indented">
                と表現できます。最後にこれを計算すると
            </p>
            <p class="history-indented hint-bgcolor-gray mono">
                &nbsp;&nbsp;%s{additionBin}<br>
                = %s{bin}<sub>(2)</sub>
            </p>
            <p class="history-indented">
                になります。
            </p>
            """

        let hint (number: int) : string =
            $"""
            <details id="hintDetails">
                <summary><h2>ヒント:</h2></summary>
                <h3>考え方 1</h3>
                %s{(newHintRepeatDivision 2 number)}
                <h3>考え方 2</h3>
                %s{(newHintRepeatAddition (Dec.Valid number) (devideIntoPowerOfTwo number))}
            </details>
            """

        let newNumberWithTwoOne min max =
            let rec newTwoRandomNumbers min max =
                let rand = new Random()
                let index1 = rand.Next(min, max)
                let index2 = rand.Next(min, max)

                if index1 <> index2 then
                    (index1, index2)
                else
                    newTwoRandomNumbers min max

            newTwoRandomNumbers min max |> Tuple.map (pown 2) ||> (+)

        let question (digit: int) (lastNumbers: int list) : int =
            newNumber (fun _ -> newNumberWithTwoOne 0 digit) (fun n -> List.contains n lastNumbers = false)

        let history (correct: bool) (input: string) : string =
            match input |> Bin.validate |> Bin.toDec with
            | Dec.Invalid _ -> ""
            | Dec.Valid v ->
                let colored = input |> Fermata.String.padLeft 8 ' ' |> escapeSpace
                let spacePadded = v |> string |> Fermata.String.padLeft 3 ' ' |> escapeSpace
                newHistory correct colored 2 spacePadded 10

        let additional number : unit =
            (document.getElementById "hint1").onclick <-
                (fun _ ->
                    (document.getElementById "hint1").innerHTML <- newHintAnimation 2 number 20
                    (document.getElementById "hintDetails").setAttribute ("open", "true"))

        let rec exec
            (questionGenerator: int list -> int)
            (hintGenerator: int -> string)
            (errorGenerator: string -> string -> exn -> string)
            (additional: int -> unit)
            (numbersToKeep: int)
            (lastNumbers: int list)
            (answer: int)
            : unit =
            // Getting the user input.
            let numberInput = document.getElementById "numberInput" :?> HTMLInputElement
            let input: string = numberInput.value |> escapeHtml
            let bin: Bin = input |> Bin.validate

            numberInput.focus ()

            match bin with
            | Bin.Invalid e ->
                // Making an error message.
                (document.getElementById "errorArea").innerHTML <- errorGenerator (string answer) input e
            | Bin.Valid v ->
                (document.getElementById "errorArea").innerHTML <- ""

                // Making a new history and updating the history with the new one.
                let outputArea = document.getElementById "outputArea" :?> HTMLParagraphElement

                match bin |> Bin.toDec with
                | Dec.Invalid _ -> ()
                | Dec.Valid d ->
                    let historyMessage =
                        history (d = answer) v
                        |> (fun x -> concatinateStrings "<br>" [ x; outputArea.innerHTML ])

                    outputArea.innerHTML <- historyMessage

                    if d <> answer then
                        ()
                    else
                        // Making the next question.
                        let nextNumber: int = questionGenerator lastNumbers
                        (document.getElementById "questionSpan").innerText <- string nextNumber
                        (document.getElementById "hintArea").innerHTML <- hintGenerator nextNumber
                        additional nextNumber

                        numberInput.value <- ""

                        // Updating `lastNumbers`.
                        // These numbers will not be used for the next question.
                        let lastNumbers' = (nextNumber :: lastNumbers) |> List.truncate numbersToKeep

                        // Setting the next answer to the check button.
                        let f =
                            fun (e: Event) ->
                                e.preventDefault ()

                                exec
                                    questionGenerator
                                    hintGenerator
                                    errorGenerator
                                    additional
                                    numbersToKeep
                                    lastNumbers'
                                    nextNumber

                        (document.getElementById "submitButton").onclick <- f
                        (document.getElementById "inputArea").onsubmit <- f

        let exec' (lastNumbers: int list) (answer: int) : unit =
            exec (question 8) hint newErrorMessageBin additional 10 lastNumbers answer

        let init'
            (questionGenerator: int list -> int)
            (hintGenerator: int -> string)
            (additional: int -> unit)
            (sourceRadix: int)
            (destinationRadix: int)
            (executor: int list -> int -> unit)
            (keyboardshortcutSetter: KeyboardEvent -> unit)
            : unit =
            // Initialization.
            let initNumber: int = questionGenerator []
            (document.getElementById "questionSpan").innerText <- string initNumber
            (document.getElementById "srcRadix").innerText <- sprintf "(%d)" sourceRadix
            (document.getElementById "dstRadix").innerText <- string destinationRadix
            (document.getElementById "binaryRadix").innerHTML <- sprintf "<sub>(%d)</sub>" destinationRadix
            (document.getElementById "hintArea").innerHTML <- hintGenerator initNumber

            let f =
                fun (e: Event) ->
                    e.preventDefault ()
                    executor [ initNumber ] initNumber

            (document.getElementById "submitButton").onclick <- f
            (document.getElementById "inputArea").onsubmit <- f

            additional initNumber

            (document.getElementById "helpButton").onclick <-
                (fun _ ->
                    [ "helpWindow"; "helpBarrier" ]
                    |> List.iter (fun x -> (document.getElementById x).classList.toggle "active" |> ignore))

            (document.getElementById "helpBarrier").onclick <-
                (fun _ ->
                    [ "helpWindow"; "helpBarrier" ]
                    |> List.iter (fun x -> (document.getElementById x).classList.remove "active" |> ignore))

            (document.getElementById "helpClose").onclick <-
                (fun _ ->
                    [ "helpWindow"; "helpBarrier" ]
                    |> List.iter (fun x -> (document.getElementById x).classList.remove "active" |> ignore))

            document.onkeydown <- (fun (e: KeyboardEvent) -> keyboardshortcutSetter e)

        let init4 () =
            init' (question 4) hint additional 10 2 exec' EndlessBinary.keyboardshortcut

        let init () : unit =
            document.title <- "10進数→2進数 (1) - taidalab"

            let header = document.querySelector "header"
            header.innerHTML <- Content.Common.header
            header.className <- "dec2bin"

            (document.getElementById "hamburgerButton").onclick <-
                (fun _ ->
                    (document.querySelector "nav").classList.toggle "flagged" |> ignore
                    (document.getElementById "barrier").classList.toggle "flagged" |> ignore
                    (document.querySelector "main").classList.toggle "flagged" |> ignore)

            (document.getElementById "barrier").onclick <-
                (fun _ ->
                    (document.querySelector "nav").classList.remove "flagged" |> ignore
                    (document.getElementById "barrier").classList.remove "flagged" |> ignore
                    (document.querySelector "main").classList.remove "flagged" |> ignore)

            (document.querySelector "#headerTitle").innerHTML <-
                """<span>10進数→2進数 (1) - </span><span translate="no">taidalab</span>"""

            (document.querySelector "main").innerHTML <- EndlessBinary.Course.main help "help-color dec2bin"
            (document.querySelector "#submitButton").className <- "dec2bin"
            (document.querySelector "#questionArea").innerHTML <- Content.Common.question

            init' (question 8) hint additional 10 2 exec' EndlessBinary.keyboardshortcut
